### ISSUE TITLE
GPU / next-gen graphics on the Web

### ISSUE THREAD

**Issue by svgeesus** [w3c/strategy#68]
See https://webkit.org/blog/7380/next-generation-3d-graphics-on-the-web/ and https://www.w3.org/community/blog/2017/02/07/proposed-group-gpu-for-the-web-community-group/

**Comment by svgeesus** [comment]
Duplicate of https://github.com/w3c/team-strat/issues/51

**Comment by sideshowbarker** [comment]
### A 3D Portability Initiative from The Khronos Group
https://lists.w3.org/Archives/Public/public-gpu/2017Mar/0000.html

> Due to industry demand, Khronos is initiating the development of a solution to enable 3D applications that are portable across Vulkan, DX12 and Metal
> A 3D Portability working group could cooperate with the SPIR-V working group to coordinate the evolution and development of open sourced shading language portability solutions, and take requirements from the needs of the next generation WebGL API for GPU graphics and compute acceleration.

-----

![various stuff feeding into WebGL](https://www.khronos.org/assets/uploads/apis/2017-3d-portability-2.png)

-----

![more arrows pointing to and from WebGL](https://cloud.githubusercontent.com/assets/194984/23452001/c990d822-fea5-11e6-818b-cf975b0f311e.png)

**Comment by svgeesus** [comment]
For a previous (failed, because pushback on maybe timing attacks and fingerprinting) Web technology that used GPU shaders: CSS custom filters
http://alteredqualia.com/css-shaders/article/
https://developers.google.com/web/updates/2013/03/Introduction-to-Custom-Filters-aka-CSS-Shaders

**Comment by dontcallmedom** [comment]
A [slideset](https://docs.google.com/presentation/d/1Z_3-3V6FRsF8OJNeH7yc6UKtgXy90Ggff07V9Z6uo6U/edit#slide=id.g2590b281ab_0_3) that describes some of the motivation for the WebGPU effort as well as some of the challenges

**Comment by plehegar** [comment]
Btw, this is useful for #28

**Comment by sideshowbarker** [comment]
Related concrete proposals:

* WebGPU API Proposal (Apple), https://github.com/w3c/strategy/issues/152
* WHLSL (Web High-level Shading Language) for WebGPU, https://github.com/w3c/strategy/issues/153
* SPIR-V for WebGPU (binary shader language), https://github.com/w3c/strategy/issues/154

**Comment by tidoust** [comment]
Advance notice about the development the draft charter sent to the Advisory Committee and [public new work mailing-list](https://lists.w3.org/Archives/Public/public-new-work/2019Nov/0002.html)

**Comment by sideshowbarker** [comment]
The GPU for Web CG formally agreed on a new shader language that they’re calling it WGSL:

https://gpuweb.github.io/gpuweb/wgsl.html

The (draft/stub) specs lists its goals as:

  * Trivially convertable to SPIR-V
  * Constructs are defined as normative references to their SPIR-V counterparts
  * All features in WGSL are directly translatable to SPIR-V. (No polymorphism, no general pointers, no overloads, etc)
  * Features and semantics are exactly the ones of SPIR-V
  * Each item in this spec must provide the mapping to SPIR-V for the construct

**Comment by tidoust** [comment]
An updated [draft charter for a GPU for the Web Working Group](https://htmlpreview.github.io/?https://github.com/grorg/admin/blob/wg-charter-draft/wg-charter.html) is available for horizontal review.

Please note that the GPU for the Web CG is still reviewing the charter and may perhaps still make further changes to it.

**Comment by michael-n-cooper** [comment]
No comments from APA. Over to @brewerj to complete accessibility horizontal review.

**Comment by wseltzer** [comment]
Another link by which to see the [draft charter](https://htmlpreview.github.io/?https://raw.githubusercontent.com/grorg/admin/wg-charter-draft/wg-charter.html)

**Comment by r12a** [comment]
Is the draft charter using an old template?  I'm looking at the 1st para in the Coordination section and not seeing what i expected.

Also, although the charter clearly indicates what it is setting out to do, and includes an explanation of what a GPU is, it doesn't seem to describe why this work is being done and what benefits will accrue from it.  It think it would be useful to add something along those lines near the top.

**Comment by himorin** [comment]
No further comment from i18n WG.

**Comment by samuelweiler** [comment]
I like that this charter requires mitigation techniques be included and gatekeeps publication on having those, but a general +1 to r12a's comment re: using the current template, at https://w3c.github.io/charter-drafts/    Start with that and add back in the good words on mitigation.

Also, I find it weird to be citing the WebPerf WG's note on security and privacy for "mitigations".  Curious as to where that came from.

That said, fingerprinting is a huge risk here, as I think was reflected in [this comment earlier in this thread](https://github.com/w3c/strategy/issues/68#issuecomment-328845111
). It seems like the charter should address that directly.  Does the WG know how it's going to gate-keep this functionality?  If not, defining that should be a named in the Scope section of the charter.  I don't know that it needs to be a separate deliverable, but it is going to be a key part of the WG's necessary work.

Another risk I see is unauthorized use of computational resources - e.g. could the GPU access be used for cryptomining?  It seems like gatekeeping will be necessary here, too.

**Comment by tidoust** [comment]
I prepared another PR with updates: https://github.com/grorg/admin/pull/2

@r12a said:
> Is the draft charter using an old template? I'm looking at the 1st para in the Coordination section and not seeing what i expected.

Yes, the initial draft charter was actually created a long time ago. I updated the first paragraph in the coordination section.


@samuelweiler said:
> Also, I find it weird to be citing the WebPerf WG's note on security and privacy for "mitigations". Curious as to where that came from.

I suspect this (and a few other occurrences) was a result of a copy and paste from the Web Performance WG charter. I removed the link in the PR (but kept the wording).

For fingerprinting and unauthorized use of computational resources, I do not know what the group has in mind. In the PR I prepared, I proposed to add work on mitigation strategies, calling out both potential issues. I didn't mention gatekeeping as that seems to be one of the strategies that could be used to mitigate the issues.

**Comment by samuelweiler** [comment]
For future reference: Mozilla has a [page about `WEBGL_debug_renderer_info`](https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_debug_renderer_info) which says:

> Depending on the privacy settings of the browser, this extension might only be available to privileged contexts or not work at all. In Firefox, if privacy.resistFingerprinting is set to true, this extensions is disabled.

Which is to say: some of the WebGL fingerprinting surface has already been recognized as a problem.

**Comment by samuelweiler** [comment]
the PR resolves my charter concerns, though the problems you're tackling here are huge.  This is going to be an exciting area to watch.

**Comment by wseltzer** [comment]
[Call for AC review](https://lists.w3.org/Archives/Public/public-new-work/2020Jun/0008.html) issued on [draft WG charter](https://www.w3.org/2020/06/gpuweb-charter.html)

**Comment by samuelweiler** [comment]
Charter approved. https://lists.w3.org/Archives/Public/public-new-work/2020Aug/0007.html
